### PR TITLE
[WIP] Add Parquet decryption support for Hive tables

### PR DESCRIPTION
**Pull request by amoghmargoor**
## Description
Adds support to read Hive tables with encrypted Parquet files. 
PS: This PR is work in progress and we are adding tests to it.

## Additional context and related issues
Parquet added support for encryption https://parquet.apache.org/docs/file-format/data-pages/encryption/. Spark also added support to read and write tables with parquet encrypted files. In this PR we are adding support to read Hive tables with encrypted Parquet files with Trino.

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by ggershinsky** [comment]
Thanks @amoghmargoor . A high-level comment first. This patch is big; I think it can be split into smaller patches that are easier to process. 
- the classes under `crypto/keytools` package handle key management. They can moved into a separate PR. This PR can focus on the basic PME layer that is given an encryption key via the reader API, and doesn't manage the key storage etc.
- many classes in the `crypto` package are identical to the same classes in the Apache parquet-java artifact. Can we re-use them?

**Comment by ggershinsky** [comment]
I think the dependency on Apache Hadoop jars is "provided" in the parquet-hadoop pom, so maybe they won't be fetched during the Trino build. Most of the basic crypto package classes (unrelated to factories) don't need Apache Hadoop artifacts. Unless there are some other cross-dependencies, we might be able to re-use those classes.

File: lib/trino-parquet/src/main/java/io/trino/parquet/EncryptionUtils.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        Optional<FileDecryptionProperties> fileDecryptionProperties = cryptoFactory.map(factory -> factory.getFileDecryptionProperties(parquetReaderOptions, filePath, trinoFileSystem));
+        return fileDecryptionProperties.map(properties -> new InternalFileDecryptor(properties));
+    }
+
+    private static Optional<TrinoDecryptionPropertiesFactory> loadDecryptionPropertiesFactory(ParquetReaderOptions trinoParquetCryptoConfig)
+    {
+        if (trinoParquetCryptoConfig.getCryptoFactoryClass() == null) {
+            return Optional.empty();
+        }
+        final Class<?> foundClass = TrinoCryptoConfigurationUtil.getClassFromConfig(
```
Comment (electrum) [inline lib/trino-parquet/src/main/java/io/trino/parquet/EncryptionUtils.java:49]: We don’t do this kind of dynamic class loading based on class names from config. Everything should be configured statically.

**Comment by electrum** [comment]
The last time I looked, the Parquet crypto classes rely on Hadoop `Configuration`. Also, they’re mixed together in a JAR with other Hadoop-requiring classes, so we don’t include it in Trino.

File: plugin/trino-hive/pom.xml [hunk 2]
```diff
@@ -379,6 +379,21 @@
             <scope>runtime</scope>
         </dependency>
 
+        <!-- Below two dependencies are not needed when parquet-kms-apple includes jackson dependency -->
```
Comment (sopel39) [inline plugin/trino-hive/pom.xml:382]: we should remove this or update the comment

**Comment by sopel39** [comment]
> The last time I looked, the Parquet crypto classes rely on Hadoop Configuration. Also, they’re mixed together in a JAR with other Hadoop-requiring classes, so we don’t include parquet-hadoop in Trino.

@electrum all crypto classes are forked into `io.trino.parquet.crypto` package. There is no dependency on `parquet-hadoop` from `trino-parquet`.

**Comment by electrum** [comment]
Sorry for the confusion. I was answering these comments:

> many classes in the crypto package are identical to the same classes in the Apache parquet-java artifact. Can we re-use them?

> Most of the basic crypto package classes (unrelated to factories) don't need Apache Hadoop artifacts. Unless there are some other cross-dependencies, we might be able to re-use those classes.

**Comment by ggershinsky** [comment]
> all crypto classes are forked into io.trino.parquet.crypto package. There is no dependency on parquet-hadoop from trino-parquet.

If no dependency on parquet-hadoop, then sure, it's fine to fork the crypto classes. They are quite stable in the Apache Parquet-java repo, have not been modified in many years (besides code style changes). The only differences in Trino will be related to the config, file path and file system objects, I'll review those parts.

File: lib/trino-parquet/src/main/java/io/trino/parquet/crypto/keytools/TrinoHadoopFSKeyMaterialStore.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+import io.trino.filesystem.TrinoFileSystem;
+import io.trino.filesystem.TrinoInputFile;
+import io.trino.filesystem.TrinoInputStream;
+import io.trino.parquet.crypto.ParquetCryptoRuntimeException;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.util.Map;
+
+public class TrinoHadoopFSKeyMaterialStore
```
Comment (ggershinsky) [inline lib/trino-parquet/src/main/java/io/trino/parquet/crypto/keytools/TrinoHadoopFSKeyMaterialStore.java:29]: might want to rename it, this class doesn't use HadoopFS

File: lib/trino-parquet/src/main/java/io/trino/parquet/ParquetReaderOptions.java [hunk 4]
```diff
@@ -118,6 +122,91 @@ public DataSize getSmallFileThreshold()
...
+    {
+        return parquetReaderEncryptionOptions.keyList;
+    }
+
+    public String getEncryptionKeyFile()
+    {
+        return parquetReaderEncryptionOptions.keyFile;
+    }
+
+    public boolean isEncryptionEnvironmentKeys()
```
Comment (ggershinsky) [inline lib/trino-parquet/src/main/java/io/trino/parquet/ParquetReaderOptions.java:195]: these are custom parameters for one particular KMS client

**Comment by ggershinsky** [comment]
> The only differences in Trino will be related to the config, file path and file system objects, I'll review those parts.

Had a look at these. The file path and file system parts are fine. 

Regarding the config (encryption parameters) - this is related to the general challenge of supporting custom KMS systems. There is basically an infinite number of key managers out there (each public cloud has one; plus open-source KMS's; plus org private KMS systems), so ideally we'd have a dynamic loading mechanism for custom KMS client classes, and a map-like config object for supplying custom encryption parameters for these custom classes.

Otherwise, the users would have to add their custom KMS client class (and perform custom modifications of the `ParquetReaderOptions.java`) in their Trino fork.

**Comment by sopel39** [comment]
I had a discussion with @electrum regarding KMS. The best would be to have one or two predefined KMS and let people contribute more to OSS or override binding if they need company specific customization

**Comment by ggershinsky** [comment]
> I had a discussion with @electrum regarding KMS. The best would be to have one or two predefined KMS and let people contribute more to OSS or override binding if they need company specific customization

SGTM

**Comment by sopel39** [comment]
Superceeded by https://github.com/trinodb/trino/pull/24517

**Pull request closed without merging**